### PR TITLE
virtualboxKvm: init

### PR DESCRIPTION
**Pull request by blitz**
## Description of changes

This PR adds support for running VirtualBox without its own kernel modules and use KVM instead. This packages the [recently published](https://cyberus-technology.de/articles/vbox-kvm-public-release) KVM patch for VirtualBox.

Still to be done:

- [x] Add a NixOS test
- [x] Fix module defaults to unbreak CI

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by blitz** [comment]
Result of `nixpkgs-review pr 288579` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
    <li>tests.nixos-functions.nixos-test</li>
  </ul>
</details>
<details>
  <summary>31 packages built:</summary>
  <ul>
    <li>linuxKernel.packages.linux_4_19.virtualbox</li>
    <li>linuxPackages_4_19_hardened.virtualbox (linuxKernel.packages.linux_4_19_hardened.virtualbox)</li>
    <li>linuxKernel.packages.linux_5_10.virtualbox</li>
    <li>linuxPackages_5_10_hardened.virtualbox (linuxKernel.packages.linux_5_10_hardened.virtualbox)</li>
    <li>linuxKernel.packages.linux_5_15.virtualbox</li>
    <li>linuxPackages_5_15_hardened.virtualbox (linuxKernel.packages.linux_5_15_hardened.virtualbox)</li>
    <li>linuxKernel.packages.linux_5_4.virtualbox</li>
    <li>linuxPackages_5_4_hardened.virtualbox (linuxKernel.packages.linux_5_4_hardened.virtualbox)</li>
    <li>linuxPackages.virtualbox (linuxKernel.packages.linux_6_1.virtualbox)</li>
    <li>linuxPackages_hardened.virtualbox (linuxPackages_6_1_hardened.virtualbox)</li>
    <li>linuxKernel.packages.linux_6_5.virtualbox</li>
    <li>linuxPackages_6_5_hardened.virtualbox (linuxKernel.packages.linux_6_5_hardened.virtualbox)</li>
    <li>linuxKernel.packages.linux_6_6.virtualbox</li>
    <li>linuxPackages_6_6_hardened.virtualbox (linuxKernel.packages.linux_6_6_hardened.virtualbox)</li>
    <li>linuxPackages_latest.virtualbox (linuxKernel.packages.linux_6_7.virtualbox)</li>
    <li>linuxPackages_6_7_hardened.virtualbox (linuxKernel.packages.linux_6_7_hardened.virtualbox)</li>
    <li>linuxPackages_latest-libre.virtualbox (linuxKernel.packages.linux_latest_libre.virtualbox)</li>
    <li>linuxPackages-libre.virtualbox (linuxKernel.packages.linux_libre.virtualbox)</li>
    <li>linuxPackages_lqx.virtualbox (linuxKernel.packages.linux_lqx.virtualbox)</li>
    <li>linuxPackages_xanmod.virtualbox (linuxKernel.packages.linux_xanmod.virtualbox)</li>
    <li>linuxPackages_xanmod_latest.virtualbox (linuxKernel.packages.linux_xanmod_latest.virtualbox ,linuxPackages_xanmod_stable.virtualbox)</li>
    <li>linuxPackages_zen.virtualbox (linuxKernel.packages.linux_zen.virtualbox)</li>
    <li>virtualbox</li>
    <li>virtualbox.modsrc</li>
    <li>virtualboxHardened</li>
    <li>virtualboxHardened.modsrc</li>
    <li>virtualboxHeadless</li>
    <li>virtualboxHeadless.modsrc</li>
    <li>virtualboxKvm</li>
    <li>virtualboxWithExtpack</li>
    <li>virtualboxWithExtpack.modsrc</li>
  </ul>
</details>

**Comment by blitz** [comment]
Result of `nixpkgs-review pr 288579` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
  </ul>
</details>
<details>
  <summary>31 packages built:</summary>
  <ul>
    <li>linuxKernel.packages.linux_4_19.virtualbox</li>
    <li>linuxKernel.packages.linux_4_19_hardened.virtualbox</li>
    <li>linuxKernel.packages.linux_5_10.virtualbox</li>
    <li>linuxKernel.packages.linux_5_10_hardened.virtualbox</li>
    <li>linuxKernel.packages.linux_5_15.virtualbox</li>
    <li>linuxKernel.packages.linux_5_15_hardened.virtualbox</li>
    <li>linuxKernel.packages.linux_5_4.virtualbox</li>
    <li>linuxKernel.packages.linux_5_4_hardened.virtualbox</li>
    <li>linuxKernel.packages.linux_6_1.virtualbox</li>
    <li>linuxKernel.packages.linux_hardened.virtualbox (linuxKernel.packages.linux_6_1_hardened.virtualbox)</li>
    <li>linuxKernel.packages.linux_6_5.virtualbox</li>
    <li>linuxKernel.packages.linux_6_5_hardened.virtualbox</li>
    <li>linuxKernel.packages.linux_6_6.virtualbox</li>
    <li>linuxKernel.packages.linux_6_6_hardened.virtualbox</li>
    <li>linuxKernel.packages.linux_6_7.virtualbox</li>
    <li>linuxKernel.packages.linux_6_7_hardened.virtualbox</li>
    <li>linuxKernel.packages.linux_latest_libre.virtualbox</li>
    <li>linuxKernel.packages.linux_libre.virtualbox</li>
    <li>linuxKernel.packages.linux_lqx.virtualbox</li>
    <li>linuxKernel.packages.linux_xanmod.virtualbox</li>
    <li>linuxKernel.packages.linux_xanmod_latest.virtualbox (linuxKernel.packages.linux_xanmod_stable.virtualbox)</li>
    <li>linuxKernel.packages.linux_zen.virtualbox</li>
    <li>virtualbox</li>
    <li>virtualbox.modsrc</li>
    <li>virtualboxHardened</li>
    <li>virtualboxHardened.modsrc</li>
    <li>virtualboxHeadless</li>
    <li>virtualboxHeadless.modsrc</li>
    <li>virtualboxKvm</li>
    <li>virtualboxWithExtpack</li>
    <li>virtualboxWithExtpack.modsrc</li>
  </ul>
</details>

**Comment by blitz** [comment]
At least @tfc has reported that the tests don't work for him. This is probably a bug or related to nested virtualization. We can consider making them optional for now. VirtualBox KVM is still pretty new and somewhat experimental in the less well-trodden paths, so issues are expected.

**Comment by blitz** [comment]
I've updated this to the latest [release](https://github.com/cyberus-technology/virtualbox-kvm/releases/tag/dev-20240226), which should improve Windows performance and also fix some issues.

**Comment by blitz** [comment]
I've also made the test optional for now. They work fine for me on a AMD and an Intel box, but as this is all a bit new, I would like to see some more people trying them out on different setups.

**Comment by blitz** [comment]
Result of `nixpkgs-review pr 288579` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
  </ul>
</details>
<details>
  <summary>1 package failed to build:</summary>
  <ul>
    <li>linuxKernel.packages.linux_libre.virtualbox</li>
  </ul>
</details>
<details>
  <summary>30 packages built:</summary>
  <ul>
    <li>linuxKernel.packages.linux_4_19.virtualbox</li>
    <li>linuxKernel.packages.linux_4_19_hardened.virtualbox</li>
    <li>linuxKernel.packages.linux_5_10.virtualbox</li>
    <li>linuxKernel.packages.linux_5_10_hardened.virtualbox</li>
    <li>linuxKernel.packages.linux_5_15.virtualbox</li>
    <li>linuxKernel.packages.linux_5_15_hardened.virtualbox</li>
    <li>linuxKernel.packages.linux_5_4.virtualbox</li>
    <li>linuxKernel.packages.linux_5_4_hardened.virtualbox</li>
    <li>linuxKernel.packages.linux_6_1.virtualbox</li>
    <li>linuxKernel.packages.linux_hardened.virtualbox (linuxKernel.packages.linux_6_1_hardened.virtualbox)</li>
    <li>linuxKernel.packages.linux_6_5.virtualbox</li>
    <li>linuxKernel.packages.linux_6_5_hardened.virtualbox</li>
    <li>linuxKernel.packages.linux_6_6.virtualbox</li>
    <li>linuxKernel.packages.linux_6_6_hardened.virtualbox</li>
    <li>linuxKernel.packages.linux_6_7.virtualbox</li>
    <li>linuxKernel.packages.linux_6_7_hardened.virtualbox</li>
    <li>linuxKernel.packages.linux_latest_libre.virtualbox</li>
    <li>linuxKernel.packages.linux_lqx.virtualbox</li>
    <li>linuxKernel.packages.linux_xanmod.virtualbox</li>
    <li>linuxKernel.packages.linux_xanmod_latest.virtualbox (linuxKernel.packages.linux_xanmod_stable.virtualbox)</li>
    <li>linuxKernel.packages.linux_zen.virtualbox</li>
    <li>virtualbox</li>
    <li>virtualbox.modsrc</li>
    <li>virtualboxHardened</li>
    <li>virtualboxHardened.modsrc</li>
    <li>virtualboxHeadless</li>
    <li>virtualboxHeadless.modsrc</li>
    <li>virtualboxKvm</li>
    <li>virtualboxWithExtpack</li>
    <li>virtualboxWithExtpack.modsrc</li>
  </ul>
</details>

**Comment by avdv** [comment]
Result of `nixpkgs-review pr 288579` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
    <li>tests.nixos-functions.nixos-test</li>
  </ul>
</details>
<details>
  <summary>29 packages built:</summary>
  <ul>
    <li>linuxKernel.packages.linux_4_19.virtualbox</li>
    <li>linuxPackages_4_19_hardened.virtualbox (linuxKernel.packages.linux_4_19_hardened.virtualbox)</li>
    <li>linuxKernel.packages.linux_5_10.virtualbox</li>
    <li>linuxPackages_5_10_hardened.virtualbox (linuxKernel.packages.linux_5_10_hardened.virtualbox)</li>
    <li>linuxKernel.packages.linux_5_15.virtualbox</li>
    <li>linuxPackages_5_15_hardened.virtualbox (linuxKernel.packages.linux_5_15_hardened.virtualbox)</li>
    <li>linuxKernel.packages.linux_5_4.virtualbox</li>
    <li>linuxPackages_5_4_hardened.virtualbox (linuxKernel.packages.linux_5_4_hardened.virtualbox)</li>
    <li>linuxKernel.packages.linux_6_1.virtualbox</li>
    <li>linuxPackages_6_1_hardened.virtualbox (linuxKernel.packages.linux_6_1_hardened.virtualbox)</li>
    <li>linuxPackages.virtualbox (linuxKernel.packages.linux_6_6.virtualbox)</li>
    <li>linuxPackages_hardened.virtualbox (linuxPackages_6_6_hardened.virtualbox)</li>
    <li>linuxPackages_latest.virtualbox (linuxKernel.packages.linux_6_7.virtualbox)</li>
    <li>linuxPackages_6_7_hardened.virtualbox (linuxKernel.packages.linux_6_7_hardened.virtualbox)</li>
    <li>linuxPackages_latest-libre.virtualbox (linuxKernel.packages.linux_latest_libre.virtualbox)</li>
    <li>linuxPackages-libre.virtualbox (linuxKernel.packages.linux_libre.virtualbox)</li>
    <li>linuxPackages_lqx.virtualbox (linuxKernel.packages.linux_lqx.virtualbox)</li>
    <li>linuxPackages_xanmod.virtualbox (linuxKernel.packages.linux_xanmod.virtualbox)</li>
    <li>linuxPackages_xanmod_latest.virtualbox (linuxKernel.packages.linux_xanmod_latest.virtualbox ,linuxPackages_xanmod_stable.virtualbox)</li>
    <li>linuxPackages_zen.virtualbox (linuxKernel.packages.linux_zen.virtualbox)</li>
    <li>virtualbox</li>
    <li>virtualbox.modsrc</li>
    <li>virtualboxHardened</li>
    <li>virtualboxHardened.modsrc</li>
    <li>virtualboxHeadless</li>
    <li>virtualboxHeadless.modsrc</li>
    <li>virtualboxKvm</li>
    <li>virtualboxWithExtpack</li>
    <li>virtualboxWithExtpack.modsrc</li>
  </ul>
</details>

**Comment by parthy** [comment]
Looks good for me too!

<details><summary>2 packages blacklisted:</summary>

- nixos-install-tools
- tests.nixos-functions.nixos-test

</details>

<details><summary>29 packages built:</summary>

- linuxKernel.packages.linux_4_19.virtualbox
- linuxPackages_4_19_hardened.virtualbox
- linuxKernel.packages.linux_5_10.virtualbox
- linuxPackages_5_10_hardened.virtualbox
- linuxKernel.packages.linux_5_15.virtualbox
- linuxPackages_5_15_hardened.virtualbox
- linuxKernel.packages.linux_5_4.virtualbox
- linuxPackages_5_4_hardened.virtualbox
- linuxKernel.packages.linux_6_1.virtualbox
- linuxPackages_6_1_hardened.virtualbox
- linuxPackages.virtualbox
- linuxPackages_hardened.virtualbox
- linuxPackages_latest.virtualbox
- linuxPackages_6_7_hardened.virtualbox
- linuxPackages_latest-libre.virtualbox
- linuxPackages-libre.virtualbox
- linuxPackages_lqx.virtualbox
- linuxPackages_xanmod.virtualbox
- linuxPackages_xanmod_latest.virtualbox
- linuxPackages_zen.virtualbox
- virtualbox
- virtualbox.modsrc
- virtualboxHardened
- virtualboxHardened.modsrc
- virtualboxHeadless
- virtualboxHeadless.modsrc
- virtualboxKvm
- virtualboxWithExtpack
- virtualboxWithExtpack.modsrc

</details>

**Comment by tfc** [comment]
Works great for me now, too.
PR looks good to me.

Would like to, but will not merge, as it feels inappropriate to merge stuff as a shareholder.

File: nixos/modules/virtualisation/virtualbox-host.nix [hunk 1]
```diff
@@ -114,17 +125,43 @@ in
...
+  } (mkIf cfg.enableKvm {
+    assertions = [
+      {
+        assertion = !cfg.addNetworkInterface;
+        message = "VirtualBox KVM only supports standard NAT networking for VMs. Please turn off virtualisation.virtualbox.host.addNetworkInferface.";
+      }
+
+      {
+        assertion = !cfg.enableHardening;
+        message = "VirtualBox KVM is not compatible with hardening: Please turn off virtualisation.virtualbox.host.enableHardening.";
```
Comment (adisbladis) [inline nixos/modules/virtualisation/virtualbox-host.nix:142]: The package itself also needs assertions for incompatibilities that are on the package level. Could you add a similar assertion to the virtualbox package?

Comment (blitz) [reply nixos/modules/virtualisation/virtualbox-host.nix:142]: I've added a similar assertion in the package. See the last commit.

**Pull request merged**